rct_image_tools/include/rct_image_tools/charuco.hpp

```diff
@@ -0,0 +1,354 @@
+/*
+By downloading, copying, installing or using the software you agree to this
+license. If you do not agree to this license, do not download, install,
+copy or use the software.
+
+                          License Agreement
+               For Open Source Computer Vision Library
+                       (3-clause BSD License)
+
+Copyright (C) 2013, OpenCV Foundation, all rights reserved.
+Third party copyrights are property of their respective owners.
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+  * Redistributions of source code must retain the above copyright notice,
+    this list of conditions and the following disclaimer.
+
+  * Redistributions in binary form must reproduce the above copyright notice,
+    this list of conditions and the following disclaimer in the documentation
+    and/or other materials provided with the distribution.
+
+  * Neither the names of the copyright holders nor the names of the contributors
+    may be used to endorse or promote products derived from this software
+    without specific prior written permission.
+
+This software is provided by the copyright holders and contributors "as is" and
+any express or implied warranties, including, but not limited to, the implied
+warranties of merchantability and fitness for a particular purpose are
+disclaimed. In no event shall copyright holders or contributors be liable for
+any direct, indirect, incidental, special, exemplary, or consequential damages
+(including, but not limited to, procurement of substitute goods or services;
+loss of use, data, or profits; or business interruption) however caused
+and on any theory of liability, whether in contract, strict liability,
+or tort (including negligence or otherwise) arising in any way out of
+the use of this software, even if advised of the possibility of such damage.
+*/
+
+#ifndef __OPENCV_CHARUCO_HPP__
+#define __OPENCV_CHARUCO_HPP__
+
+#include <opencv2/core.hpp>
+#include <vector>
+#include <opencv2/aruco.hpp>
```

Comment:
This looks like a file that might already be in OpenCV. If it is, let's just include it directly from OpenCV rather than copy-pasting it here

Reply:
It's part of OpenCV 3.0 or greater, so is it better to copy paste it?

Reply:
ROS Kinetic and later depend on OpenCV 3.X, so as long as we're building this package on Kinetic or newer we'll always have a comparatively up-to-date version of OpenCV that includes the 3rd-party contrib packages which provide the ArUco library. This specific header is available through `#include <opencv2/aruco/charuco.hpp`.

One challenge is that the Canonical-sourced OpenCV package for Ubuntu 16.04 (Xenial) is actually OpenCV 2.X. For a potential application where `rct_image_tools` is being used in a non-ROS context on a Xenial system, we'd have to specifically require OpenCV 3 in our `CMakeLists.txt`. This would require the user to manually provide that version of the library rather than being able to rely on `apt install` to provide the right version. I don't think this is a use case we need to specifically support, since as currently designed the higher-level packages in this repo like `rct_examples` depend on ROS (and by the time we'll probably get around to making the whole repo ROS-agnostic it'll be April 2021 and Xenial will finally be out of support).

As a side node OpenCV's file naming conventions are kind of weird. I think they did a big refactor between 1.X and 2.X and changed the directory of their headers to use `opencv2`. However, 3.X uses the same directory, so there isn't an `opencv3` directory.